wasm/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+#![feature(alloc_error_handler)]
+// Required because of `unsafe` code and `no_mangle` use
+#![allow(unsafe_code)]
+#![no_std]
+
+extern crate alloc;
+
+use alloc::{boxed::Box, format, vec::Vec};
+
+use data_model::prelude::*;
+pub use iroha_data_model as data_model;
+pub use iroha_wasm_derive::iroha_wasm;
+use parity_scale_codec::{Decode, Encode};
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+#[cfg(target_pointer_width = "32")]
+type WasmUsize = u32;
+#[cfg(target_pointer_width = "64")]
+type WasmUsize = u64;
```

Comment:
Maybe only switch based on `cfg(test)`?

Reply:
No. I think that this is precisely what it needs to be. Anything else, e.g. `cfg(test)` relies on some assumptions that may not be true in the future.

Reply:
no, no, this is **target**`_pointer_width`, not host. It's the pointer size for the wasm target which is, at this moment still, 32-bit (`wasm32-unknown-unknown`). It is possible that wasm will support 64-bit targets in the future as the specification progresses.

Primary reason for this was because of tests which, if compiled for the entire workspace, are compiled for 64-bit architecture and start failing if 32-bit target architecture is expected. Of course tests can be compiled for 32-bit architecture for this crate specifically but that adds up some friction for the dev.

Also I think I'm going to have to add another CI job for building this crate, I don't think it can be part of the workspace because it introduces it's own panic handler which results in some conflict

Reply:
I think I may have misunderstood your original comment. 

I agree that a separate CI job is good. Actually a separate workflow, since everything related to WASM is in a few localised paths.

I'm not sure that your assertion about this being only useful for tests is going to hold true in the future. However, for now adding `#[cfg(test)]` is fine, and I doubt that 64-bit pointers are going to make it into WASM any time soon. I don't know if adding an extra layer of complexity via `#[cfg(test)]` is worth the few microseconds of debug build time, but it's up to you.

Reply:
With the `compile_error`s added, I think this discussion can be closed